src/hotspot/share/classfile/javaClasses.cpp

```diff
@@ -2780,6 +2780,26 @@ Handle java_lang_Throwable::get_cause_with_stack_trace(Handle throwable, TRAPS)
   return h_cause;
 }
 
+Handle java_lang_Throwable::get_cause_simple(JavaThread* current, Handle throwable) {
```

Comment:
How about `get_cause_without_stacktrace`?

The two methods should be able to share a lot of code rather than duplicating most of it.

Reply:
That would be one option. There is a lot of overlap with these two methods. Also I think your new method has to handle this case too:
```
// If new_exception returns a different exception while creating the exception, return null.
```

Reply:
>  Also I think your new method has to handle this case too
> / If new_exception returns a different exception while creating the exception, return null.

I'm afraid that in this case ( if I return null ), the original cause again will be missed from the stacktrace.

Reply:
If `new_exception` returns a different exception then you have already lost that. The `ExceptionInInitializerError` would have been replaced by some other exception object: most likely OOME or even another SOE?

Reply:
Is it poor, to report OOME or SOE as the cause, instead of `ExceptionInInitializerError`?  
Especially when the first call to `get_cause_with_stack_trace` already has returned null.
Actually, this is the intention of the fix - to report the SOE as the cause of NCDFE.

Reply:
If an exception happens during clinit then it is caught and set as the cause of an ExceptionInInitializerError and the class is marked as erroneous. Subsequent attempts to load the class should throw NCDFE with the EIIE as the cause.

Sorry I've now confused myself about exactly what the problem scenario is and how we are trying to fix it. The first attempt to call `get_cause_with_stack_trace` fails due to SOE - but what "cause" are we trying to get here and what exception are we trying to throw? I need to take a deeper look at this again tomorrow and get the control flow sorted out.